src/celeritas/track/SimParams.cc

```diff
@@ -0,0 +1,88 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/track/SimParams.cc
+//---------------------------------------------------------------------------//
+#include "SimParams.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/data/CollectionBuilder.hh"
+#include "celeritas/field/FieldPropagator.hh"
+#include "celeritas/io/ImportData.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/track/SimData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with imported data.
+ */
+std::shared_ptr<SimParams>
+SimParams::from_import(ImportData const& data, SPConstParticles particle_params)
+{
+    CELER_EXPECT(data);
+    CELER_EXPECT(particle_params);
+    CELER_EXPECT(data.trans_params.looping.size() == particle_params->size());
+
+    SimParams::Input input;
+    input.particles = std::move(particle_params);
+
+    // Calculate the maximum number of steps a track below the threshold energy
+    // can take while looping (ceil(max Geant4 field propagator substeps / max
+    // Celeritas field propagator substeps))
+    auto max_substeps = FieldPropagatorOptions::max_substeps;
+    CELER_ASSERT(data.trans_params.max_substeps >= max_substeps);
+    auto max_subthreshold_steps
+        = (data.trans_params.max_substeps + (max_substeps - 1)) / max_substeps;
```

Comment:
using `corecel/math/Algorithms.hh`:
```suggestion
    auto max_subthreshold_steps
        = ceil_div(data.trans_params.max_substeps, max_substeps);
```

Reply:
Perfect, forgot we had that :)